pkg/connector/user.go

```diff
@@ -45,40 +46,185 @@ func userResource(ctx context.Context, user *client.ConfluenceUser) (*v2.Resourc
 		return nil, err
 	}
 
-	return resource, nil
+	return newUserResource, nil
 }
 
-func (o *userResourceType) List(ctx context.Context, _ *v2.ResourceId, pt *pagination.Token) ([]*v2.Resource, string, annotations.Annotations, error) {
-	l := ctxzap.Extract(ctx)
+// parsePageToken given a marshalled pageToken as a string, return the pageToken
+// bag and the current page number.
+func parsePageToken(
+	pageToken string,
+	resourceID *v2.ResourceId,
+) (
+	*pagination.Bag,
+	string,
+	error,
+) {
+	b := &pagination.Bag{}
+	err := b.Unmarshal(pageToken)
+	if err != nil {
+		return nil, "0", err
+	}
+
+	if b.Current() == nil {
+		b.Push(pagination.PageState{
+			ResourceTypeID: resourceID.ResourceType,
+			ResourceID:     resourceID.Resource,
+		})
+	}
 
-	users, _, err := o.client.GetUsers(ctx, "", ResourcesPageSize)
+	page := b.PageToken()
+	return b, page, nil
+}
+
+func (o *userResourceType) List(
+	ctx context.Context,
+	_ *v2.ResourceId,
+	pToken *pagination.Token,
+) (
+	[]*v2.Resource,
+	string,
+	annotations.Annotations,
+	error,
+) {
+	logger := ctxzap.Extract(ctx)
+	logger.Debug("Starting Users List", zap.String("token", pToken.Token))
+
+	// There is no Confluence Cloud REST API to get all users, so get all groups
+	// and then all members of each group.
+
+	// The second parameter here is "user", which acts as a default value.
+	bag, page, err := parsePageToken(
+		pToken.Token,
+		&v2.ResourceId{ResourceType: resourceTypeUser.Id},
+	)
 	if err != nil {
 		return nil, "", nil, err
 	}
-	rv := make([]*v2.Resource, 0)
-	for _, user := range users {
-		if user.AccountType != accountTypeAtlassian {
-			l.Debug("confluence: user is not of type atlassian", zap.Any("user", user))
-			continue
+
+	outputResources := make([]*v2.Resource, 0)
+	var outputAnnotations annotations.Annotations
+	switch bag.ResourceTypeID() {
+	case resourceTypeUser.Id:
+		logger.Debug("Got a user from the bag", zap.String("page", page))
+		if page == "" {
+			page = "0"
+		}
+		// Add a new page of groups
+		groups, nextToken, ratelimitData, err := o.client.GetGroups(
+			ctx,
+			page,
+			ResourcesPageSize,
+		)
+		logger.Debug(
+			"Got groups",
+			zap.Int("len", len(groups)),
+			zap.String("nextToken", nextToken),
+		)
+		outputAnnotations = WithRateLimitAnnotations(ratelimitData)
+		if err != nil {
+			return nil, "", outputAnnotations, err
+		}
+
+		// Push next page to stack. (Short-circuits if token is "".)
+		err = bag.Next(nextToken)
+		if err != nil {
+			return nil, "", outputAnnotations, err
 		}
 
-		userCopy := user
-		ur, err := userResource(ctx, &userCopy)
+		for _, group := range groups {
+			logger.Debug(
+				"adding a group to the bag",
+				zap.String("id", group.Id),
+				zap.String("name", group.Name),
+			)
+			bag.Push(
```

Comment:
I think this will exceed the max page token size of 4096 characters if the customer has a lot of groups.

Reply:
Should I bring down the page size when getting the list of groups? Is there a way to make sure that there is only one page worth of groups in the bag at a time?

Reply:
Hard to say. Also I'm pretty sure it's possible to have a user that isn't a member of any group.